### PR TITLE
Buffer dropped messages stats

### PR DESCRIPTION
**Pull request by hansieodendaal**
## Description
Added logging for buffer dropped messages stats and a subscriber id for identification. Breaking change.

## Motivation and Context
Monitoring is needed for dropped messages in the pub_sub subscription channels.

## How Has This Been Tested?
Performed a system-level test on Windows as part of a Tari base node executable.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [X] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [X] New Tests
* [ ] Documentation

## Checklist:

* [X] I ran `cargo-fmt --all` before pushing.
* [X] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [X] I have added tests to cover my changes.

### CODE REVIEW

**Review by philipr-za — DISMISSED** [review]

File: Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,8 @@ edition = "2018"
 arc-swap = "0.4.2"
 futures = { version = "^0.3.1" }
 crossbeam-channel = "0.3.9"
+log = "0.4.6"
+log4rs = {version = "0.8.3", features = ["console_appender", "file_appender", "file", "yaml_format"]}
```
Comment (CjS77) [inline Cargo.toml]: This is a library, so you should drop this dependency

File: src/async_channel.rs [hunk 2]
```diff
@@ -62,17 +73,40 @@ impl<T> Stream for Subscriber<T> {
     fn poll_next(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
         self.sleeper.register(cx.waker());
         match self.receiver.try_recv() {
-            Ok(item) => Poll::Ready(Some(item)),
-            Err(error) => match error {
-                TryRecvError::Empty => Poll::Pending,
-                TryRecvError::Disconnected => Poll::Ready(None),
+            Ok(item) => {
+                if self.receiver.get_dropped_messages_state() {
+                    error!(
```
Comment (CjS77) [inline src/async_channel.rs]: Suggest making with lower - I'd suggest trace or debug

File: src/async_channel.rs [hunk 3]
```diff
@@ -5,11 +5,22 @@ use futures::{
     Sink,
     Stream,
 };
+use log::*;
 use std::{pin::Pin, sync::Arc, task::Poll};
 
-pub fn bounded<T>(size: usize) -> (Publisher<T>, Subscriber<T>) {
-    let (sender, receiver) = raw_bounded(size);
+const LOG_TARGET: &str = "pub_sub";
```
Comment (CjS77) [inline src/async_channel.rs]:
```suggestion
const LOG_TARGET: &str = "tari_broadcast_channel::async_channel";
```

LOG_TARGET should direct you to the source file

File: src/channel.rs [hunk 4]
```diff
@@ -122,16 +134,37 @@ impl<T> Receiver<T> {
...
+                        .set(self.messages_dropped_count.get() + self.ri.get() - temp_ri);
+                    self.messages_dropped_state.store(true, Ordering::Relaxed);
+                }
                 self.ri.inc();
                 return Ok(val);
             }
         }
     }
+
+    /// Returns the amount of dropped messages
```
Comment (CjS77) [inline src/channel.rs]:
```suggestion
    /// Returns the total number of dropped messages for the entire history of this receiver
```

File: src/channel.rs [hunk 5]
```diff
@@ -122,16 +134,37 @@ impl<T> Receiver<T> {
...
+    pub fn get_dropped_messages_count(&self) -> usize {
+        self.messages_dropped_count.get()
+    }
+
+    /// Returns the dropped message state when the last message was read
```
Comment (CjS77) [inline src/channel.rs]:
```suggestion
    /// Indicates whether messages were dropped the last time the message buffer was read
```

File: src/channel.rs [hunk 6]
```diff
@@ -145,6 +178,9 @@ impl<T> Clone for Receiver<T> {
             size: self.size,
             sub_count: Arc::clone(&self.sub_count),
             is_sender_available: self.is_sender_available.clone(),
+            messages_dropped_count: AtomicCounter::new(self.messages_dropped_count.get()),
```
Comment (CjS77) [inline src/channel.rs]: Wouldn't we want this to be zero? This new receiver will have it's own statistics so why would you want to inherit the number of dropped messages?

**Review by CjS77 — COMMENTED** [review]
Looks fine; a few nits though

File: src/channel.rs (hunk 6 above)
Comment (hansieodendaal) [reply src/channel.rs]: Yes, good point. A new subscriber starts fresh.

File: src/async_channel.rs (hunk 2 above)
Comment (hansieodendaal) [reply src/async_channel.rs]: Going with `debug`, as `trace` has way too much noise, thousands of empty buffer read attempts and cloning of subscribers.

**Review by CjS77 — DISMISSED** [review]
LGTM

**Review by CjS77 — APPROVED** [review]

**Pull request merged**